### PR TITLE
Issue #13501: Kill mutation for CodePointUtil

### PR DESCRIPTION
**Pull request by Kevin222004**
Issue #13501: Kill mutation for CodePointUtil

Mutation introduction at https://github.com/checkstyle/checkstyle/commit/b8628532f12f46ab18e61372a997a4ca73e33d2a
@nrmancuso  is mandatory reviewer, as we allowed this code in PR year ago.

------

# Mutations
https://github.com/checkstyle/checkstyle/blob/bb64c20cbcf6e543b297b6e809f379bd8746b5e7/config/pitest-suppressions/pitest-utils-suppressions.xml#L174-L226

-----

# Explaintaion 

Both the method 
https://github.com/checkstyle/checkstyle/blob/79e760900c400c0ebac121a67517235529012eb3/src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java#L64-L70
and
https://github.com/checkstyle/checkstyle/blob/79e760900c400c0ebac121a67517235529012eb3/src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java#L78-L86
are same only logic has written differently.
in `stripTrailing` method loop start from lastIndex of array and finsih at index 0 and same but opposite in `stripLeading`
the condition of loop is `isCodePointWhitespace(codePoints, startIndex)` will fail in both the method if their is no whitespace

------

# Regression 

Report-1 :- https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/2237062_2023171438/reports/diff/index.html

Report-2 :- https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/2237062_2023180212/reports/diff/index.html

Report-3 :- https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/2237062_2023191958/reports/diff/index.html

Report-4 :- 

Report-5 :-

------

Diff Regression config: https://gist.githubusercontent.com/Kevin222004/66e91526d84b11d76ed35dc61f651108/raw/01d1012ebdb5cae259e1f1775c9b25d68c41c881/codeutil2.xml
Diff Regression projects: https://gist.githubusercontent.com/Kevin222004/2207ddd139a1721d00b0e6b312bc3f03/raw/a0b63175830fccda7c242058d8f513d61d4dce0d/util1.properties
Report label: Aug-28-2.2

### CODE REVIEW

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
@romani please re-run pitest-common

**Comment by romani** [comment]
Relaunched

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java [hunk 1]
```diff
@@ -56,45 +56,19 @@ public static boolean hasWhitespaceBefore(int index, int... codePoints) {
...
 
     /**
-     * Removes trailing whitespaces.
+     * Removes whitespaces.
      *
      * @param codePoints array of unicode code points
      * @return unicode code points array with trailing whitespaces removed
      */
-    public static int[] stripTrailing(int... codePoints) {
+    public static int[] stripWhitespace(int... codePoints) {
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java:64]: The name is confusing. The logic still strips only trailing whitespaces. Why did you rename it?

Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/CodePointUtil.java]: yes, you are right. this logic strips only trailing whitespace. I have change the name again

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Maybe I am missing some context here, but I have some questions before we even dive into changing any code here or considering any code point specifics:
1. Where is the hardcoded mutation + report?
2. Why can't we find input to cover each mutation?

**Review by rdiachenko — APPROVED** [review]
lgtm

Either `stripLeading` or `stripTrailing` will work here, `trim` is redundant. That's because of these conditions: `substringBeforeToken.length == 0` and `substringAfterToken.length == 0`. They mean that there is only a token text on the current line which may be surrounded by whitespaces. Those whitespaces are stipped be either of the above function.

@Kevin222004 please correct me if I'm wrong.

**Comment by Kevin222004** [comment]
@nrmancuso https://github.com/checkstyle/checkstyle/pull/13473#pullrequestreview-1587409884

> Where is the hardcoded mutation + report? 

This code is not such big so I have directly evaluated this code. I have never generated report for all mutation after hardcode it 

> Why can't we find input to cover each mutation?

https://github.com/checkstyle/checkstyle/pull/13473#pullrequestreview-1587676680 @rdiachenko I agree.
@nrmancuso The whole overview is if there is whitespace after or before the token text then it becomes 0 else it wouldn't and to check this condition we don't require trim only a single method is enough to handle this either strip trailing or strip leading because here we only care about whitespace.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/SeparatorWrapCheck.java [hunk 2]
```diff
@@ -224,10 +224,10 @@ public void visitToken(DetailAST ast) {
         final int colNo = ast.getColumnNo();
         final int lineNo = ast.getLineNo();
         final int[] currentLine = getLineCodePoints(lineNo - 1);
-        final int[] substringAfterToken = CodePointUtil.trim(
+        final int[] substringAfterToken = CodePointUtil.stripTrailing(
                 Arrays.copyOfRange(currentLine, colNo + text.length(), currentLine.length)
         );
-        final int[] substringBeforeToken = CodePointUtil.trim(
+        final int[] substringBeforeToken = CodePointUtil.stripTrailing(
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/SeparatorWrapCheck.java:230]: `trim` was easier to understand, but `stripTrailing` semantics are not obvious here, especially with before/after naming.  Can we just use CodePointUtil#isBlank?

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Item:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/SeparatorWrapCheck.java (hunk 2 above)
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/SeparatorWrapCheck.java]: Can we just use CodePointUtil#isBlank?

yes, please check I have updated

**Comment by Kevin222004** [comment]
Github, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/SeparatorWrapCheck.java [hunk 3]
```diff
@@ -224,19 +224,19 @@ public void visitToken(DetailAST ast) {
         final int colNo = ast.getColumnNo();
         final int lineNo = ast.getLineNo();
         final int[] currentLine = getLineCodePoints(lineNo - 1);
-        final int[] substringAfterToken = CodePointUtil.trim(
+        final boolean isTokenEmptyAfterSubstring = CodePointUtil.isBlank(
                 Arrays.copyOfRange(currentLine, colNo + text.length(), currentLine.length)
         );
-        final int[] substringBeforeToken = CodePointUtil.trim(
+        final boolean isTokenEmptyBeforeSubstring = CodePointUtil.isBlank(
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/SeparatorWrapCheck.java:230]: `isTokenEmptyBeforeSubstring` -> `isLineEmptyBeforeToken`

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Please rearrange declaration order and order of usage to be the same, ex:
```java
boolean x = ...;
boolean y = ...;

if (x) {
...
} else if (y) {
...
}

```

Items:

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by nrmancuso** [comment]
@Kevin222004 please make CI happy

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/SeparatorWrapCheck.java [hunk 4]
```diff
@@ -224,19 +224,19 @@ public void visitToken(DetailAST ast) {
         final int colNo = ast.getColumnNo();
         final int lineNo = ast.getLineNo();
         final int[] currentLine = getLineCodePoints(lineNo - 1);
-        final int[] substringAfterToken = CodePointUtil.trim(
+        final boolean isTokenEmptyAfterSubstring = CodePointUtil.isBlank(
```
[the remark this answers is not in the document]
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/SeparatorWrapCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/SeparatorWrapCheck.java (hunk 3 above)
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/SeparatorWrapCheck.java]: done

**Comment by romani** [comment]
@Kevin222004 , please generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Review by nrmancuso — APPROVED** [review]

**Comment by Kevin222004** [comment]
Github, generate report

**Review by romani — APPROVED** [review]
Ok to merge
